### PR TITLE
Implement mutate pass from pass spec RFC

### PR DESCRIPTION
**Pull request by PoignardAzur**
This is part of the "Pass Specification" RFC:
https://github.com/linebender/rfcs/pull/7

Rename WidgetCtx to MutateCtx.
Add a mutate pass.
Add a `mutate_later` context method to trigger that pass.
Refactor `edit_root_widget` to use a version of that pass.
Add a separate constructor for the synthetic WidgetState created in RenderRoot.

### CODE REVIEW

**Comment by PoignardAzur** [comment]
Note: this was split off from #498 and rebased after #488 was merged.

**Comment by PoignardAzur** [comment]
I think this is ready to merge.

File: masonry/src/contexts.rs
Comment (waywardmonkeys) [inline masonry/src/contexts.rs:496]: Would be good to have a doc comment for this method and the one below. Probably a pretty short one.

File: masonry/src/render_root.rs
Comment (waywardmonkeys) [inline masonry/src/render_root.rs:348]: If mutations are required to several widgets, is there an inefficiency here? (Question, not looking for a change, I think.)

Comment (PoignardAzur) [reply masonry/src/render_root.rs:348]: If you mutate several widgets with the same call to `edit_widget`, no. If you call `edit_widget` hundreds of times per frame, yes, a bit.

It's an escape hatch first and foremost, it's not supposed to substitute for the pass system.

File: masonry/src/passes/mutate.rs
Comment (Philipp-M) [inline masonry/src/passes/mutate.rs:26]: Yeah good question, I think it depends on whether there should be "dangling" widgets, in which case an `Option` certainly makes sense. But when not, I think this is fine, would reduce boilerplate everywhere else.

File: masonry/src/render_root.rs
Comment (Philipp-M) [reply masonry/src/render_root.rs:348]: So when I'm getting it right, later only `edit_root_widget` and `mutate_later` should be used?

Is this method even necessary (outside of tests)? I only see usage of it in the tests, so should this be `pub(crate)` instead?

File: masonry/src/widget/widget_state.rs
Comment (Philipp-M) [inline masonry/src/widget/widget_state.rs:154]: I wonder why this was not necessary before, or was this a bug, that was just never surfaced? (xilem "classic" had this too btw. with the Changeflags)

File: masonry/src/widget/widget_state.rs
Comment (Philipp-M) [inline masonry/src/widget/widget_state.rs:156]: should this also be:
```suggestion
            request_anim: true,
```
?
I see that this is set in the `Spinner` widget at `WidgetAdded`. Or should this explicitly be set somewhere else for widgets needing animation directly on creation (like the `Spinner`)?

**Review by Philipp-M — APPROVED** [review]
I wonder if we slowly want to remove the `get_raw_mut` stuff? (I think that is planned isn't it?)

(I agree with Bruce, I think the methods could have some doc-comments ideally when it makes sense with doc-tests, but I don't want to block on that, when you add these later, it would just also help with reviewing.
Thanks for keeping this PR more focused, that makes reviewing easier and more effective)

Thanks looks good apart from a few more open questions.

File: masonry/src/passes/mutate.rs
Comment (DJMcNab) [inline masonry/src/passes/mutate.rs:26]: I think the move to `Option` should happen in this PR. AIUI, the only thing this impacts is the `Drop` implementation, right? So it should be a relatively undisruptive change.

File: masonry/src/passes/mutate.rs
Comment (DJMcNab) [inline masonry/src/passes/mutate.rs:19]: Would it be worth this being
```suggestion
    let _span = info_span!("mutate_widget", name = state_mut.short_name).entered();
```
(or whatever the actually correct name is)

File: masonry/src/passes/mutate.rs
Comment (DJMcNab) [inline masonry/src/passes/mutate.rs:36]:
```suggestion

// Merge any state changes up to the root.
```

File: masonry/src/passes/mutate.rs
Comment (DJMcNab) [inline masonry/src/passes/mutate.rs:45]:
```suggestion
/// Apply any deferred mutations (created using [...Ctx::mutate_later](crate::LayoutCtx::mutate_later)).
pub(crate) fn run_mutate_pass(root: &mut RenderRoot, root_state: &mut WidgetState) {
```

File: masonry/src/contexts.rs
Comment (DJMcNab) [inline masonry/src/contexts.rs:491]: Are these meaningful on `MutateCtx`?

File: masonry/src/passes/mutate.rs
Comment (DJMcNab) [inline masonry/src/passes/mutate.rs:15]:
```suggestion
pub(crate) fn mutate_widget<R>(
    root: &mut RenderRoot,
    id: WidgetId,
    mutate_fn: impl FnOnce(WidgetMut<'_, Box<dyn Widget>>) -> R,
) -> R {
```

File: masonry/src/widget/widget_state.rs
Comment (DJMcNab) [inline masonry/src/widget/widget_state.rs:173]: I find this `root` thing quite confusing, because my general understanding is that the root widget has its own state.

**Review by DJMcNab — APPROVED** [review]
A clear step towards the required results.

File: masonry/src/render_root.rs
Comment (PoignardAzur) [reply masonry/src/render_root.rs:348]: @waywardmonkeys requested it for his use-cases.

It's not necessary for Xilem, but it's easy to implement and I can easily imagine use-cases for it for consumers of Masonry.

File: masonry/src/passes/mutate.rs
Comment (PoignardAzur) [reply masonry/src/passes/mutate.rs:26]: Yeah, it's a whole thing.

Other reason to have parent state as an Option would be for reborrowed WidgetMut, to replaced the `is_reborrow` flag.

File: masonry/src/widget/widget_state.rs
Comment (PoignardAzur) [reply masonry/src/widget/widget_state.rs:154]: The default constructor in xilem classic's `ChangeFlags` would set these to true.

The way it worked in Druid and Masonry is that the default constructor set flags used by the composition root (eg `Window` for Druid, `RenderRoot` for Masonry) and then the `WidgetAdded` event set a bunch of flags to true.

File: masonry/src/passes/mutate.rs
Comment (PoignardAzur) [reply masonry/src/passes/mutate.rs:19]: I guess that makes sense, since where your "root" is would be pretty important information in that case.

File: masonry/src/widget/widget_state.rs
Comment (PoignardAzur) [reply masonry/src/widget/widget_state.rs:173]: It's useful for lifetime purposes; but yeah, I think at this point it can probably be considered vestigial. I might remove it once other passes are implemented.

File: masonry/src/contexts.rs
Comment (PoignardAzur) [reply masonry/src/contexts.rs:491]: Maybe.

Either way, I'm trying to consider these methods less in terms of "does this context absolutely need it" and more "does it break something if this context has it". I think the context APIs are easier to reason about that way.

File: masonry/src/widget/widget_state.rs
Comment (PoignardAzur) [reply masonry/src/widget/widget_state.rs:156]: No, I think you're right. Having all the machinery run once by default is easier. Fewer footguns for users that way.

**Comment by PoignardAzur** [comment]
I think I've addressed all reviews. Will merge this tonight if nobody raises objections.

File: masonry/examples/calc_masonry.rs
Comment (DJMcNab) [inline masonry/examples/calc_masonry.rs:149]: I think that this should have a comment, explaining why we can't access the background/box directly.

File: masonry/src/contexts.rs
Comment (DJMcNab) [inline masonry/src/contexts.rs:382]: Maybe something like this would be clearer.
```suggestion
            // We don't don't reborrow `parent_widget_state`. This avoids running
            // `merge_up` in `Drop` multiple times for this widget (it will still be called when
            // `self` is dropped)
```

File: masonry/src/passes/mutate.rs
Comment (DJMcNab) [inline masonry/src/passes/mutate.rs:57]: We should reuse the allocation here (comment on the implicit `drop(callbacks)`)

Of course, we need to be aware of the fact that new callbacks could have been added, but that won't be the general case.

File: masonry/src/contexts.rs
Comment (DJMcNab) [reply masonry/src/contexts.rs:493]: It makes reuse of the `mutate_callbacks` vector potentially fraught, and the semantics aren't especially clear.

Effectively, if you're calling this, you should be using a different method on the `MutateCtx` which you already have...

File: masonry/src/render_root.rs
Comment (DJMcNab) [inline masonry/src/render_root.rs:373]: It's not clear to me what role we expect `post_event_processing` to have "long-term".

File: masonry/src/render_root.rs
Comment (DJMcNab) [inline masonry/src/render_root.rs:355]:
```suggestion
        let res = mutate_widget(self, id, f);
```

**Review by DJMcNab — COMMENTED** [review]
A few more details, nothing architecturally problematic

File: masonry/src/render_root.rs
Comment (PoignardAzur) [reply masonry/src/render_root.rs:373]: I'll remove it once all the passes are split out from RenderRoot and WidgetPod.

File: masonry/src/passes/mutate.rs
Comment (PoignardAzur) [reply masonry/src/passes/mutate.rs:57]: Hard disagree. Skipping one allocation per frame is not worth even one line of code. Especially if we need additional logic in case new callbacks have been added (though I understand you think that shouldn't be possible).

Comment (DJMcNab) [reply masonry/src/passes/mutate.rs:57]: It's not necessarily just one allocation, if this feature is used widely. But sure, I'm not going to block on this.

File: masonry/src/contexts.rs
Comment (Philipp-M) [reply masonry/src/contexts.rs:493]: Agree, I don't see benefits of this yet (other than being confusing, or allure users to "hacky" workarounds). I think we should see how far we get without it on `MutateCtx`. (I'm specifically also talking about `mutate_callbacks` not sure about the rest, that may be useful)

File: masonry/src/passes/mutate.rs
Comment (Philipp-M) [reply masonry/src/passes/mutate.rs:57]: Yeah likely more allocations, in case there are added a few callbacks (which seems likely). I don't yet see how your case applies, it's sync isn't it? It could also be double-buffered with a second `Vec` and `std::mem::swap` alternatively. But yeah also no strong feelings here (just seems to be a cheap optimization, which can still be reversed later)...

**Review by Philipp-M — APPROVED** [review]
Apart from the two comments seems good to land.

File: masonry/src/passes/mutate.rs
Comment (PoignardAzur) [reply masonry/src/passes/mutate.rs:57]: Right.

Even then, `mutate_callbacks` is a Vec of boxes, so we're talking about saving `O(log(N))` reallocs per frame in a context where we have `O(N)` allocations minimum. There's no situation in which that meaningfully impacts performance.

And I think `render_root` is *the* file where you can to keep complexity down as much as you can. It's a complexity bottleneck, so to speak, both because it naturally tends to amass more code as features get added, and because anyone who wants to global view of the code needs to read it.

File: masonry/src/contexts.rs
Comment (PoignardAzur) [reply masonry/src/contexts.rs:493]: Will keep as-is for now, with a TODO comment. We can reconsider when passes are all implemented.

**Pull request merged**
